Shift object and exit waves during centring probe

ptim0626 opened the pull request:
Previously the object and exit waves are not shifted when the probe is shifted, resulting in incorrect results. After shifting the object and exit waves with the same amount of offset, it produces correct result.

ptim0626 committed:
Shift object and exit waves during centring probe

bjoernenders commented on ptypy/engines/projectional.py:
```diff
@@ -314,17 +314,25 @@ def overlap_update(self):
+                ex_s = pr_s.views[0].pod.ex_view.storage
+                ex_s.data[:] = u.shift_zoom(ex_s.data, (1.,)*3,
+                        (0, c1[0], c1[1]), (0, c2[0], c2[1]))
 
```
Probably doesn't work with the BlockModel. We would need to account for multiple exit wave storages to be linked to one probe storage

ptim0626 committed:
[WIP] mass center cuda

ptim0626 committed:
Finished mass_center for 2D

ptim0626 committed:
Completed the mass_center 3D case

ptim0626 committed:
Loop through different exit wave storages when centering probe

ptim0626 committed:
Fix a bug in the size of threads and block in final_sums in mass_center

daurer commented:
@ptim0626 could you also implement the modified `center_probe` feature for `ptypy/engines/stochastic.py`?

ptim0626 committed:
Simplify the looping syntax

ptim0626 committed:
Implement center_probe in stochastic algorithm

ptim0626 committed:
Put center_probe in projectional_serial

ptim0626 committed:
Finish the abs2sum kernel

ptim0626 committed:
[WIP] interpolated shift kernel

The starting value of a block for positive shift is wrong, and the last value
of a block for negative shift is wrong??

ptim0626 committed:
Fix a bug in linear_interpolate_kernel in interpolated shift

The four corners of the Halo were missing, leading to the usage of random values
in shared memory when performing linear interpolation. The four corners of the
Halo are now defined correctly.

ptim0626 committed:
Fix another bug in linear_interpolate_kernel (swapped rows and columns)

rows and columns variables are swapped, may result in pre-mature return.

ptim0626 committed:
Finish the interpolated shift kernel

ptim0626 committed:
Implement center_probe in projectional_pycuda

ptim0626 committed:
Reduce number of loopings when shifting exit waves

daurer commented on ptypy/engines/projectional.py:
```diff
@@ -329,10 +329,10 @@ def center_probe(self):
+                # shift the exit waves
+                for ex_s in self.ex.storages.values():
+                    ex_s.data[:] = u.shift_zoom(ex_s.data, (1.,)*3,
+                            (0, c1[0], c1[1]), (0, c2[0], c2[1]))
```
@ptim0626 there is a problem here when there are multiple storages (multiple scans). Since we are iterating over the probe storages already, we should only shift the exit wave storage that is associated with the probe storage. Unfortunately, the `ID` attribute of the storages differs between probe and exit, but we could potentially use the `numID` attribute to fix this, something like this:
```
for ex_s in self.ex.storages.values():
    if ex_s.numID == pr_s.numID:
        ex_s.data[:] = u.shift_zoom(ex_s.data, (1.,)*3,
                   (0, c1[0], c1[1]), (0, c2[0], c2[1]))
```

daurer replied:
As discussed, unfortunately this approach also does not work. Best is to revert back to the previous solution iterating through all the views.

daurer commented:
As a final change, we should move the center_probe call out of the overlap update and place it directly in the main iterate loop between overlap and position updates...

ptim0626 committed:
Revert "Reduce number of loopings when shifting exit waves"

This reverts commit 4e5e22cd39ba887a7aad44ae1e47d53fa09ed6c6.

ptim0626 committed:
Adjust the way to access object and gpu data in projectional

ptim0626 committed:
Move the center_probe method out of inner loop in projectional

ptim0626 committed:
Adjust the way to access object in stochastic

ptim0626 committed:
Move the center_probe method out of inner loop in stochastic

ptim0626 committed:
Implement the center_probe in stochastic pycuda
